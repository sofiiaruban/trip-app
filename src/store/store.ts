import { configureStore } from '@reduxjs/toolkit'
import { setupListeners } from '@reduxjs/toolkit/query'
import { weatherDataApi } from '../services/weatherDataApi'
import tripReducer from './tripSlice'
import { persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import { combineReducers } from 'redux'

const persistConfig = {
  key: 'root',
  version: 1,
  storage
}

const reducer = combineReducers({
  [weatherDataApi.reducerPath]: weatherDataApi.reducer,
  trips: tripReducer
})
const persistedReducer = persistReducer(persistConfig, reducer)

export const store = configureStore({
  reducer: persistedReducer,

  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }).concat(weatherDataApi.middleware)
})

setupListeners(store.dispatch)
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
