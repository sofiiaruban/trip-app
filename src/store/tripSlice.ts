import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Trip } from '@app/types/types'
import initialTrips from '@app/initialTrips'

const initialState: Trip[] = initialTrips

const tripSlice = createSlice({
  name: 'trip',
  initialState,
  reducers: {
    addTrip(state, action: PayloadAction<Trip>) {
      state.push(action.payload)
    }
  }
})

export const { addTrip } = tripSlice.actions
export default tripSlice.reducer
