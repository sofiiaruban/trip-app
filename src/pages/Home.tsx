import { FC, Suspense, lazy, useEffect, useState } from 'react'
import { MAIN_PAGE_TEST_ID } from './constants'
import styles from './Home.module.css'
import { useSelector } from 'react-redux'
import { RootState } from '@app/store/store'
import filterTripsByCityName from '@app/helpers/filterTripsByCityName'
import sortTripsByStartDate from '@app/helpers/sortTripsByDate'
import useActiveIndex from '@app/components/TripsList/useActiveIndex'
import {
  useGetWeatherDateRangeQuery,
  useGetWeatherTodayQuery
} from '@app/services/weatherDataApi'
import getYearMonthDayDate from '@app/helpers/getYearMonthDayDate'
import { ForecastCardType } from '@app/components/ForecastList/constants'
import CountdownComponent from '@app/components/Countdown/Countdown'

const Search = lazy(() => import('@app/components/Search/Search'))
const TripsList = lazy(() => import('@app/components/TripsList/TripsList'))
const ForecastList = lazy(
  () => import('@app/components/ForecastList/ForecastList')
)
const ForecastCard = lazy(
  () => import('@app/components/ForecastList/ForecastCard/ForecastCard')
)

const Home: FC = () => {
  const selectTrips = (state: RootState) => state.trips
  const trips = useSelector(selectTrips)
  const [filteredTrips, setFilteredTrips] = useState(trips)
  const { activeIndex, setActive } = useActiveIndex()

  useEffect(() => {
    setFilteredTrips(trips)
  }, [trips])

  const onSearch: (value: string) => void = (value) => {
    const filteredList = filterTripsByCityName(trips, value)
    const sortedList = sortTripsByStartDate(filteredList)
    setFilteredTrips(sortedList)
  }

  const { data, error, isLoading } = useGetWeatherDateRangeQuery({
    cityName: filteredTrips[activeIndex].cityName,
    fromDate: getYearMonthDayDate(filteredTrips[activeIndex].startDate),
    toDate: getYearMonthDayDate(filteredTrips[activeIndex].endDate)
  })

  const {
    data: dailyWeather,
    error: isError,
    isLoading: isLoader
  } = useGetWeatherTodayQuery(filteredTrips[activeIndex].cityName)

  return (
    <div className={styles.container}>
      <main data-testid={MAIN_PAGE_TEST_ID} className={styles.home}>
        <p className={styles.mainTitle}>
          Weather <b>Forecast</b>
        </p>
        <Suspense fallback={<div>Loading Search...</div>}>
          <Search placeholder="Search your trip" onSearch={onSearch} />
        </Suspense>
        <Suspense fallback={<div>Loading TripsList...</div>}>
          <TripsList
            trips={filteredTrips}
            activeIndex={activeIndex}
            setActive={setActive}
          />
        </Suspense>
        <p className={styles.title}>Week</p>
        <Suspense fallback={<div>Loading ForecastList...</div>}>
          {error ? (
            <>Oh no, there was an error</>
          ) : isLoading ? (
            <>Loading...</>
          ) : data ? (
            <ForecastList forecastList={data?.days} />
          ) : null}
        </Suspense>
      </main>
      <aside className={styles.aside}>
        <Suspense fallback={<div>Loading ForecastCard...</div>}>
          {isError ? (
            <>Oh no, there was an error</>
          ) : isLoader ? (
            <>Loading...</>
          ) : dailyWeather ? (
            <ForecastCard
              date={dailyWeather.days[0].datetime}
              iconSrc={`trip-app/src/assets/${dailyWeather.days[0].icon}.svg`}
              iconName={dailyWeather.days[0].icon}
              degree={`${dailyWeather.days[0].temp}`}
              type={ForecastCardType.BIG}
            >
              <p className={styles.cityTitle}>{dailyWeather.address}</p>
              <CountdownComponent
                date={new Date(filteredTrips[activeIndex].startDate)}
              />
            </ForecastCard>
          ) : null}
        </Suspense>
      </aside>
    </div>
  )
}

export default Home
