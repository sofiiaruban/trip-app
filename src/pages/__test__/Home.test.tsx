import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import Home from '../Home'
import { Provider } from 'react-redux'
import { store } from '@app/store/store'

jest.mock('@reduxjs/toolkit/query/react', () => ({
  ...jest.requireActual('@reduxjs/toolkit/query/react'),
  fetchBaseQuery: jest.fn()
}))

test('Renders the main page', () => {
  const { container } = render(
    <Provider store={store}>
      <Home />
    </Provider>
  )
  expect(container).toMatchSnapshot()
})
