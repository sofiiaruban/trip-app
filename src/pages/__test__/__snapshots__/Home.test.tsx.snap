// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the main page 1`] = `
<div>
  <main
    class="home"
    data-testid="mainPage"
  >
    <div>
      <input
        class="input"
        placeholder="Search your trip"
        type="text"
        value=""
      />
      <div
        class="searchButtonIcon"
      >
        <button
          class="buttonIcon"
          type="button"
        >
          <svg
            class="searchIcon"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M18.319 14.4326C20.7628 11.2941 20.542 6.75347 17.6569 3.86829C14.5327 0.744098 9.46734 0.744098 6.34315 3.86829C3.21895 6.99249 3.21895 12.0578 6.34315 15.182C9.22833 18.0672 13.769 18.2879 16.9075 15.8442C16.921 15.8595 16.9351 15.8745 16.9497 15.8891L21.1924 20.1317C21.5829 20.5223 22.2161 20.5223 22.6066 20.1317C22.9971 19.7412 22.9971 19.1081 22.6066 18.7175L18.364 14.4749C18.3493 14.4603 18.3343 14.4462 18.319 14.4326ZM16.2426 5.28251C18.5858 7.62565 18.5858 11.4246 16.2426 13.7678C13.8995 16.1109 10.1005 16.1109 7.75736 13.7678C5.41421 11.4246 5.41421 7.62565 7.75736 5.28251C10.1005 2.93936 13.8995 2.93936 16.2426 5.28251Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      class="tripsList"
    >
      <li>
        <article
          class="trip"
        >
          <img
            alt="New York city"
            class="cityImg"
            src="https://github.com/sofiiaruban/delivery-app/assets/37212452/d5c5a07f-e0e3-4731-aec2-905b550f0137"
          />
          <div
            class="tripInfo"
          >
            <p
              class="cityName"
            >
              New York
            </p>
            <p
              class="tripDates"
            >
              <span>
                3/1/2024
              </span>
               - 
              <span>
                3/5/2024
              </span>
            </p>
          </div>
        </article>
      </li>
      <li>
        <article
          class="trip"
        >
          <img
            alt="London city"
            class="cityImg"
            src="https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a"
          />
          <div
            class="tripInfo"
          >
            <p
              class="cityName"
            >
              London
            </p>
            <p
              class="tripDates"
            >
              <span>
                4/10/2024
              </span>
               - 
              <span>
                4/15/2024
              </span>
            </p>
          </div>
        </article>
      </li>
      <li>
        <article
          class="trip"
        >
          <img
            alt="London city"
            class="cityImg"
            src="https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a"
          />
          <div
            class="tripInfo"
          >
            <p
              class="cityName"
            >
              London
            </p>
            <p
              class="tripDates"
            >
              <span>
                4/23/2024
              </span>
               - 
              <span>
                4/29/2024
              </span>
            </p>
          </div>
        </article>
      </li>
      <li>
        <article
          class="trip"
        >
          <img
            alt="London city"
            class="cityImg"
            src="https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a"
          />
          <div
            class="tripInfo"
          >
            <p
              class="cityName"
            >
              London
            </p>
            <p
              class="tripDates"
            >
              <span>
                4/18/2024
              </span>
               - 
              <span>
                4/25/2024
              </span>
            </p>
          </div>
        </article>
      </li>
      <li>
        <article
          class="trip"
        >
          <img
            alt="Kyiv city"
            class="cityImg"
            src="https://github.com/sofiiaruban/delivery-app/assets/37212452/22af8aca-39f2-481a-b092-84921b709f2a"
          />
          <div
            class="tripInfo"
          >
            <p
              class="cityName"
            >
              Kyiv
            </p>
            <p
              class="tripDates"
            >
              <span>
                5/20/2024
              </span>
               - 
              <span>
                5/25/2024
              </span>
            </p>
          </div>
        </article>
      </li>
      <li
        class="addTripItem"
      >
        <button
          class="buttonIcon"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            />
          </svg>
          <p
            class="addTripTitle"
          >
            Add trip
          </p>
        </button>
      </li>
    </ul>
    [{"cityImgSrc":"https://github.com/sofiiaruban/delivery-app/assets/37212452/d5c5a07f-e0e3-4731-aec2-905b550f0137","cityName":"New York","startDate":"2024-03-01T00:00:00.000Z","endDate":"2024-03-05T00:00:00.000Z"},{"cityImgSrc":"https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a","cityName":"London","startDate":"2024-04-10T00:00:00.000Z","endDate":"2024-04-15T00:00:00.000Z"},{"cityImgSrc":"https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a","cityName":"London","startDate":"2024-04-23T00:00:00.000Z","endDate":"2024-04-29T00:00:00.000Z"},{"cityImgSrc":"https://github.com/sofiiaruban/delivery-app/assets/37212452/6ddc7f4c-8b48-4146-890f-fc5124892c7a","cityName":"London","startDate":"2024-04-18T00:00:00.000Z","endDate":"2024-04-25T00:00:00.000Z"},{"cityImgSrc":"https://github.com/sofiiaruban/delivery-app/assets/37212452/22af8aca-39f2-481a-b092-84921b709f2a","cityName":"Kyiv","startDate":"2024-05-20T00:00:00.000Z","endDate":"2024-05-25T00:00:00.000Z"}]
  </main>
</div>
`;
